moz_security/sandboxes/heka/analysis/moz_security_pathsetfreq.lua

```diff
@@ -0,0 +1,174 @@
+
+-- send_iprepd = false -- optional, if true plugin will generate iprepd violation messages
+-- violation_type = "fxa:client_error_rate" -- required in violations mode, iprepd violation type
+set_threshold = 50 -- consecutive repeats to generate violation/tsv output
+acceptable_variance = 5 -- >= unique request paths to ignore client for interval
+-- no_single_set = true -- if true, disable tracking set with single path (e.g., repeated)
+```
+--]]
+
+require "string"
+require "table"
+
+local uri = require "lpeg.uri"
+
+local id_field          = read_config("id_field") or error("id_field must be configured")
+local id_fieldc         = read_config("id_field_capture")
+local request_field     = read_config("request_field") or error("request_field must be configured")
+local request_fieldc    = read_config("request_field_capture")
+local set_threshold     = read_config("set_threshold") or error("set_threshold must be configured")
+local accept_variance   = read_config("acceptable_variance") or error("acceptable_variance must be configured")
+local no_single         = read_config("no_single_set")
+
+local tbsend
+local violation_type
+local send_iprepd = read_config("send_iprepd")
+if send_iprepd then
+    tbsend = require "heka.iprepd".send
+    violation_type = read_config("violation_type") or error("violation_type must be configured")
+end
+
+local list = {}
+
+function process_message()
+    local id = read_message(id_field)
+    if not id then return -1, "no id_field" end
+    if id_fieldc then
+        id = string.match(id, id_fieldc)
+        if not id then return 0 end -- no error as the capture may intentionally reject entries
+    end
+    local req = read_message(request_field)
+    if not req then return -1, "no request_field" end
+    if request_fieldc then
+        req = string.match(req, request_fieldc)
+        if not req then return 0 end -- no error as the capture may intentionally reject entries
+    end
+
+    local m = uri.uri_reference:match(req)
+    if not m or not m.path then return 0 end
+
+    if not list[id] then
+        list[id] = {
+            paths   = { m.path },   -- path set list
+            r       = 0,            -- consecutive occurrences in stream
+            varied  = false,        -- true if acceptable variance
+            estab   = false,        -- true if set established
+            setptr  = 0             -- index into path list for established entry
+        }
+        return 0
+    end
+    local w = list[id]
+
+    if w.varied then return 0 end -- acceptable variance, ignore for remainder of interval
+
+    local pathsetlen = #w.paths
+    local foundi = 0
+    for i,v in ipairs(w.paths) do
+        if v == m.path then
+            foundi = i
+            break
+        end
+    end
+
+    if not w.estab then
+        if foundi == 0 then -- path isn't known
+            if pathsetlen + 1 >= accept_variance then -- addition meets variance requirement
+                w.varied = true
```

Comment:
We may want to dump the path table when we no longer need it to get it GC'ed a bit quicker (below also).